keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,442 @@
+- Resource overcommitment, fractional resource requirements, native compute resource (i.e., cpu and memory) with special metadata requirements, and group compute resources. They are out of our current scope.
+
+## Components
+### ResourceClass API
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying node level compute resource plus metadata matching. We can extend this API to better support cluster resources and group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // raw resource name. E.g.: nvidia.com/gpu
+        ResourceName string
+        // ResourceSelector  selects resources. ORed from each selector.
+        resourceSelector []ResourcePropertySelector
+}
+
+type ResourcePropertySelector struct {
+        // A list of resource/device selector requirements
+        MatchExpressions []ResourceSelectorRequirement
+}
+
+// A resource selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values
+type ResourceSelectorRequirement struct {
+        // The label key that the selector applies to
+        Key string
+        // Example 0.1, intel etc
+        Values []string
+        // Similar to NodeSelectorOperator. Valid operators are: "In", "NotIn",
+        // "Exists", "DoesNotExist", "Gt", and "Lt".
+        Operator ResourceSelectorOperator
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  resourceName: "nvidia.com/gpu"
+  resourceSelector:
+    - matchExpressions:
+        - key: "memory"
+          operator: "Gt"
+          values:
+            - "15G"
+
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  containers:
+    - name: example-container
+      resources:
+        limits:
+          nvidia.high.mem: 2
+
+```
+Above resource class will select all the nvidia-gpus which have memory greater than 15GB.
+
+YAML example 2:
+```yaml
+kind: ResourceClass
+metadata:
+  name: fast.nic
+spec:
+  resourceName: "sfc.com/smartNIC"
+  labelSelector:
+    - matchExpressions:
+        - key: "speed"
+          operator: "Gt"
+          values:
+            - "40GBPS"
```

Comment:
Also... "40GBPS" is not a simple number, how do you propose handling units wrt comparisons?  What entity has to understand the units?

Reply:
it will be similar to how millicore units and memory units are handled in existing code already.

Reply:
@vikaschoudhary16 Right... but *some* entity has to understand the units.  I imagine there are many many kinds of units one might support.  What entity has to understand these units?  Effectively we are implicitly introducing a 'type' here by adding units... where before in the Device Plugin API we had an int.  I'm just curious how new 'types' get added... and how we handle collisions of unit abbreviations.